src/util/color/color.h

```diff
@@ -0,0 +1,55 @@
+#ifndef PREDEFINEDCOLORSET_H
+#define PREDEFINEDCOLORSET_H
+
+#include <QColor>
+#include <QHash>
+
+#include "predefinedcolorsset.h"
+#include "util/math.h"
+
+namespace Color {
+    static const PredefinedColorsSet predefinedColorSet = PredefinedColorsSet();
+
+    // algorithm by http://www.nbdtech.com/Blog/archive/2008/04/27/Calculating-the-Perceived-Brightness-of-a-Color.aspx
+    // NOTE(Swiftb0y): please suggest if I should you use other methods
+    // (like the W3C algorithm) or if this approach is to to performance hungry
+    // NOTE: the author did not take alpha transparency into account!
+    static inline int brightness(int red, int green, int blue) {
+        return static_cast<int>(sqrtf(
+            red * red * .241 +
+            green * green * .691 +
+            blue * blue * .068)
+        );
+    };
+
+    static inline int brightness(const QColor& color) {
+        return brightness(color.red(), color.green(), color.red());
+    }
+
+    static inline bool isDimmColor(const QColor& color) {
+        return brightness(color) <= 127;
+    }
+
+    // If the colorToChooseBy is darker than the global threshold,
+    // dimmColor will be returned. Otherwise brightColor will be returned.
+    static inline QColor chooseColorByBrightness(QColor colorToChooseBy, QColor dimmColor , QColor brightColor) {
+        return isDimmColor(colorToChooseBy) ? dimmColor : brightColor;
+    }
+
+    // If the baseColor is darker than the global threshold,
+    // returns white, otherwise returns black.
+    static inline QColor chooseContrastColor(QColor baseColor) {
+        QColor lightColor = baseColor.lighter();
+        QColor darkColor = baseColor.darker();
+
+        // QColor::lighter() multiplies the HSV Value by some factor. When baseColor is black, Value is 0,
+        // thus after multiplication it's still 0 and we get the same color.
+        // We manually set lightColor to darkGray in this case.
+        if (baseColor.toHsv().value() == 0) {
```

Comment:
I think we should check here if the darker colour is black, else the difference between the dark and the base colour might be to small.

Reply:
Like `baseColor.rgb() == 0x0`?
Yes maybe the Hsv conversion was over-complicated.

Reply:
No, the HSV conversion is OK. value() (V of HSV) is the brightness. If this is already near 0, plain black dos not look different. 

See the implementation of darker: 
https://github.com/qt/qtbase/blob/5733dfbd90fd059e7310786faefb022b00289592/src/gui/painting/qcolor.cpp#L2740
It just does V = V / 2 for darker and
V = V * 1.5 for lighter. 

I think we should inline the lighter() and darker() code to get around multiple Hsv conversions. 

How about to use 5 % V as minimum difference before using the transparent trick for darker and brighter?

Reply:
> I think we should inline the lighter() and darker() code to get around multiple Hsv conversions.

Sorry I don't get what you mean. But maybe we can just precalculate the lightness value once and Save it on PredefinedColor.

> How about to use 5 % V as minimum difference before using the transparent trick for darker and brighter?

Do you mean the trick to use darkGray as the brighter color for really dark colors? Yeah it's good idea to use a threshold and not only use this trick for pure black.

Reply:
Added the threshold.

Reply:
> Sorry I don't get what you mean. But maybe we can just precalculate the lightness value once and Save it on PredefinedColor.

It is just that we currently use three times toHsv() in a row. If we inline Qt's code of darker and lighter we get around it. The only thing that we have to do for it is to V = V / 2 for darker and
V = V * 1.5 for lighter.

Reply:
Aaaa ok. You are right. The [lighter() code](https://code.woboq.org/qt5/qtbase/src/gui/painting/qcolor.cpp.html#_ZNK6QColor5lightEi) is not completely trivial however.

But if the color is already converted to hsl, it avoids the conversion. So if we convert the color before all the code in this method, we will only convert once. Is this enough?

Reply:
Yes, any optimization works for me.

Reply:
Done